bytecomp/translcore.ml

```diff
@@ -1234,8 +1234,10 @@ and transl_setinstvar self var expr =
 
 and transl_record env all_labels repres lbl_expr_list opt_init_expr =
   let size = Array.length all_labels in
-  (* Determine if there are "enough" new fields *)
-  if 3 + 2 * List.length lbl_expr_list >= size
+  (* Determine if there are "enough" fields (only relevant if this is a
+     functional-style record update *)
+  let no_init = match opt_init_expr with None -> true | _ -> false in
+  if no_init || size >= Config.max_young_wosize
```

Comment:
The direction of this test is wrong: should be size <= Config.max_young_wosize

Reply:
this will cause code bloat in pathological cases

Reply:
What cases do you have in mind? If the size is of the record is larger than 256 then it uses the "duplicate and modify" approach which requires less code.